### PR TITLE
Fix up bignum support.

### PR DESCRIPTION
**Pull request by aardvark179**
Adds support for relevant bignum cext functions, and alters the spec to take account of full range fixnums, and to add more thorough testing of rb_big_pack.

### CODE REVIEW

File: truffleruby/src/main/c/cext/ruby.c [hunk 1]
```diff
@@ -700,43 +700,49 @@ VALUE rb_integer_unpack(const void *words, size_t numwords, size_t wordsize, siz
...
 }
 
 unsigned long rb_big2ulong(VALUE x) {
-  rb_tr_error("rb_big2ulong not implemented");
+  return truffle_invoke_l(RUBY_CEXT, "rb_num2ulong", x);
 }
 
 VALUE rb_big_cmp(VALUE x, VALUE y) {
-  rb_tr_error("rb_big_cmp not implemented");
+  return (VALUE) truffle_invoke((void *)x, "<=>", y);
```
Comment (chrisseaton) [inline truffleruby/src/main/c/cext/ruby.c]: You shouldn't need to cast to `void *`.

File: truffleruby/src/main/java/org/truffleruby/cext/CExtNodes.java [hunk 2]
```diff
@@ -250,6 +251,56 @@ public Object ulong2num(long num) {
...
+                    }
+                }
+                System.out.flush();
+                longs[limit] = x;
+            }
+            if (negative) {
+                for (int i = blank_start; i < length; i++) {
+                    longs[i] = -1;
+                }
+            }
```
Comment (eregon) [inline truffleruby/src/main/java/org/truffleruby/cext/CExtNodes.java]: This seems very complex logic. Could it be simplified for instance by enforcing that the length given is the proper one?
What is `fraction`?
I would have thought this is easier written in Ruby, where you could extract the longs with % and / and then pack/unpack to change endianness if required.

File: truffleruby/src/main/java/org/truffleruby/cext/CExtNodes.java [hunk 3]
```diff
@@ -250,6 +251,56 @@ public Object ulong2num(long num) {
...
+                for (int i = 0; i < fraction; i++) {
+                    long a_byte = bytes[bytes.length - 1 - limit * 8 - i] & 0xffl;
+                    x = x | (a_byte << (i * 8));
+                }
+                if (negative) {
+                    for (int i = fraction; i < 8; i++) {
+                        x = x | (0xffl << (i * 8));
+                    }
+                }
+                System.out.flush();
```
Comment (eregon) [inline truffleruby/src/main/java/org/truffleruby/cext/CExtNodes.java]: Debug code

File: spec/ruby/optional/capi/bignum_spec.rb [hunk 4]
```diff
@@ -7,18 +7,29 @@ def ensure_bignum(n)
...
+
 describe "CApiBignumSpecs" do
   before :each do
     @s = CApiBignumSpecs.new
-    @max_long = ensure_bignum(2**(0.size * 8 - 1) - 1)
-    @min_long = ensure_bignum(-@max_long - 1)
-    @max_ulong = ensure_bignum(2**(0.size * 8) - 1)
-  end
+
+    if full_range_longs
```
Comment (eregon) [inline spec/ruby/optional/capi/bignum_spec.rb]: Could this condition be done with `fixnum_max` and `fixnum_min`? If so we don't need to maintain an explicit list and duplicate logic.

File: spec/ruby/optional/capi/bignum_spec.rb [hunk 5]
```diff
@@ -7,18 +7,29 @@ def ensure_bignum(n)
...
+      @min_long = ensure_bignum(-@max_long - 1)
+      @max_ulong = ensure_bignum(2**(0.size * 8) - 1)
+    end
+end
 
   describe "rb_big2long" do
-    it "converts a Bignum" do
-      @s.rb_big2long(@max_long).should == @max_long
-      @s.rb_big2long(@min_long).should == @min_long
+    if !full_range_longs
```
Comment (eregon) [inline spec/ruby/optional/capi/bignum_spec.rb]: `unless`

File: spec/ruby/optional/capi/bignum_spec.rb [hunk 6]
```diff
@@ -117,6 +117,67 @@ def ensure_bignum(n)
...
+      platform_is wordsize: 64 do
+        it "packs max_ulong into _two_ ulongs to allow sign bit" do
+          val = @s.rb_big_pack_length(@max_ulong)
+          val.should == 2
+          val = @s.rb_big_pack_array(@max_ulong, 2)
+          val[0].should == @max_ulong
+          val[1].should == 0
+        end
+
+        it "packs A 72 bit positive Bignum into 2 unsigned longs" do
```
Comment (eregon) [inline spec/ruby/optional/capi/bignum_spec.rb]: `packs a 72-bit`

File: spec/ruby/optional/capi/bignum_spec.rb [hunk 7]
```diff
@@ -117,6 +117,67 @@ def ensure_bignum(n)
...
+        end
+
+        it "packs A 2 ** 71 + 1 positive Bignum into 2 unsigned longs of 1 and 0x80" do
+          num = 2 ** 71 + 1
+          val = @s.rb_big_pack_array(num, 2)
+          val[0].should == 1;
+          val[1].should == 0x80;
+        end
+
+        it "packs A 2 ** 71 + 1 bit negative Bignum into 2 unsigned longs of 1 and ulongmax - 0xff + 0x80" do
```
Comment (eregon) [inline spec/ruby/optional/capi/bignum_spec.rb]: Avoid too much computations in the spec description, also it does not match what's in the code below.

File: spec/ruby/optional/capi/bignum_spec.rb [hunk 8]
```diff
@@ -117,6 +117,67 @@ def ensure_bignum(n)
...
+          num = -(2 ** 71 + 1)
+          val = @s.rb_big_pack_array(num, @s.rb_big_pack_length(num))
+          val[0].should == @max_ulong;
+          val[1].should == @max_ulong - 0x80;
+        end
+
+        it "packs lower order bytes into lsbs of longs for positive bignum" do
```
Comment (eregon) [inline spec/ruby/optional/capi/bignum_spec.rb]: What is `lsbs`?

File: truffleruby/src/main/java/org/truffleruby/cext/CExtNodes.java [hunk 9]
```diff
@@ -250,6 +251,56 @@ public Object ulong2num(long num) {
 
     }
 
+    @CoreMethod(names = "rb_big_pack", isModuleFunction = true, required = 2)
+    public abstract static class BIGPACKNode extends CoreMethodArrayArgumentsNode {
```
Comment (aardvark179) [inline truffleruby/src/main/java/org/truffleruby/cext/CExtNodes.java]: Do you want me to change all the other class names in CExtNodes to camelcase?

Comment (eregon) [inline truffleruby/src/main/java/org/truffleruby/cext/CExtNodes.java]: Yes, unless it's already a macro full-upperscase like `NUM2INT`.

Comment (eregon) [inline truffleruby/src/main/java/org/truffleruby/cext/CExtNodes.java]: All the existing nodes previous to this PR seem to follow CamelCase from a quick look.

File: spec/ruby/optional/capi/bignum_spec.rb (hunk 4 above)
Comment (aardvark179) [inline spec/ruby/optional/capi/bignum_spec.rb]: I've changed it to do that at the top of the file.

File: truffleruby/src/main/java/org/truffleruby/cext/CExtNodes.java (hunk 2 above)
Comment (aardvark179) [inline truffleruby/src/main/java/org/truffleruby/cext/CExtNodes.java]: I've added a comment. This area is likely to get refactored soon anyway as it is a special case of rb_integer_pack.

Comment (chrisseaton) [inline truffleruby/src/main/java/org/truffleruby/cext/CExtNodes.java]: > Is there plans to implement rb_integer_pack in the near future?

Yes we should do that too.

File: truffleruby/src/main/java/org/truffleruby/cext/CExtNodes.java [hunk 10]
```diff
@@ -235,6 +253,107 @@ public Object ulong2num(long num) {
...
+    @CoreMethod(names = "rb_big_pack", isModuleFunction = true, required = 2)
+    public abstract static class BigPackNode extends CoreMethodArrayArgumentsNode {
+
+        // The Ruby MRI API for extracting the contents of a Bignum
+        // fills a provided buffer of longs with a two's compliment
+        // representation of the number starting with the least
+        // significant part. If the provided buffer and length are
+        // longer than required for this then the remaining longs will
+        // be filled in so as to preserve the number in two's
+        // compliment form.
```
Comment (eregon) [inline truffleruby/src/main/java/org/truffleruby/cext/CExtNodes.java]: complement

File: truffleruby/src/main/java/org/truffleruby/cext/CExtNodes.java [hunk 11]
```diff
@@ -235,6 +253,107 @@ public Object ulong2num(long num) {
...
+        //
+        // 2. Pack any remaining bytes into a final long and pad it
+        // appropriately to preserve the sign of the number.
+        //
+        // 3. Fill any remaining space in the requested buffer size
+        // with longs (again preserving the sign of the number).
+        //
+        // It is important we correctly fill buffers which are longer
+        // than requested both for compatibility with MRI and to
+        // preserve the returned number.
```
Comment (eregon) [inline truffleruby/src/main/java/org/truffleruby/cext/CExtNodes.java]: That helps, thanks.

**Pull request merged**